prompt_toolkit/layout/processors.py

```diff
@@ -219,7 +219,7 @@ def apply_transformation(self, transformation_input):
                 # visualise the selection.
                 return Transformation([(selected_fragment, ' ')])
             else:
-                for i in range(from_, to + 1):
+                for i in range(from_, to):
```

Comment:
I think this change should not be here. If I undo it, the last character is selected again. 

As I noted in the issue, the visual selection (in master) matches what is selected just fine.  The issue is just what gets selected, particularly when a selection is started.